terraform/transform_targets.go

```diff
@@ -91,123 +67,63 @@ func (t *TargetsTransformer) selectTargetedNodes(g *Graph, addrs []addrs.Targeta
 			}
 		}
 	}
-	return t.addDependencies(targetedNodes, g)
-}
-
-func (t *TargetsTransformer) addDependencies(targetedNodes dag.Set, g *Graph) (dag.Set, error) {
-	// Handle nodes that need to be included if their dependencies are included.
-	// This requires multiple passes since we need to catch transitive
-	// dependencies if and only if they are via other nodes that also
-	// support TargetDownstream. For example:
-	// output -> output -> targeted-resource: both outputs need to be targeted
-	// output -> non-targeted-resource -> targeted-resource: output not targeted
-	//
-	// We'll keep looping until we stop targeting more nodes.
-	queue := targetedNodes.List()
-	for len(queue) > 0 {
-		vertices := queue
-		queue = nil // ready to append for next iteration if neccessary
-		for _, v := range vertices {
-			// providers don't cause transitive dependencies, so don't target
-			// downstream from them.
-			if _, ok := v.(GraphNodeProvider); ok {
-				continue
-			}
 
-			dependers := g.UpEdges(v)
-			if dependers == nil {
-				// indicates that there are no up edges for this node, so
-				// we have nothing to do here.
-				continue
-			}
+	// It is expected that outputs which are only derived from targeted
+	// resources are also updated. While we don't include any other possible
+	// side effects from the targeted nodes, these are added because outputs
+	// cannot be targeted on their own.
+	// Start by finding the root module output nodes themselves
+	for _, v := range vertices {
+		// outputs are all temporary value types
+		tv, ok := v.(graphNodeTemporaryValue)
+		if !ok {
+			continue
+		}
 
-			dependers = dependers.Filter(func(dv interface{}) bool {
-				_, ok := dv.(GraphNodeTargetDownstream)
-				return ok
-			})
+		// root module outputs indicate that while they are an output type,
+		// they not temporary and will return false here.
+		if tv.temporaryValue() {
+			continue
+		}
 
-			if dependers.Len() == 0 {
+		// If this output is descended only from targeted resources, then we
+		// will keep it
+		deps, _ := g.Ancestors(v)
+		found := 0
+		for _, d := range deps {
+			switch d.(type) {
+			case GraphNodeResourceInstance:
+			case GraphNodeConfigResource:
+			default:
 				continue
 			}
 
-			for _, dv := range dependers {
-				if targetedNodes.Include(dv) {
-					// Already present, so nothing to do
-					continue
-				}
-
-				// We'll give the node some information about what it's
-				// depending on in case that informs its decision about whether
-				// it is safe to be targeted.
-				deps := g.DownEdges(v)
-
-				depsTargeted := deps.Intersection(targetedNodes)
-				depsUntargeted := deps.Difference(depsTargeted)
-
-				if dv.(GraphNodeTargetDownstream).TargetDownstream(depsTargeted, depsUntargeted) {
-					targetedNodes.Add(dv)
-					// Need to visit this node on the next pass to see if it
-					// has any transitive dependers.
-					queue = append(queue, dv)
-				}
+			if !targetedNodes.Include(d) {
```

Comment:
My intuition for this, based on how we framed this mechanism when originally implementing it, is that it would include any out put that has depends on _at least one_ resource that is targeted, not that it would require _all_ of its dependencies to be targeted.

For example, if the output's expression were `a + b` and `b` were targeted, then applying that change could potentially change the value of `b` and thus require the output to be updated to show the correct result, even though `a` isn't targeted and thus can't be changing.

I suppose we probably already broke that behavior as part of making output errors always be shown and never ignored, because otherwise `a + b` could fail if `b` is targeted and `a` doesn't exist yet. If it was already broken in 0.12 then no problem with keeping it broken in the same way for now (we can decide in a later release what this behavior _ought_ to be). I'm raising this here just to confirm that it isn't a behavior change relative to 0.12, regardless of what the behavior had been in earlier versions.

Reply:
Yes, we had to remove that when outputs started producing errors. The `filterPartialOutputs` function (now removed) attempted to do what this new code does, and removed outputs from the graph when they reference nodes that aren't targeted. Whatever course of action taken on outputs that depends on targeted and un-targeted resources is, it will undoubtedly be wrong in some sense. 

Since the point of targeting is also to not include side effects of the targeted resources, I feel like adding outputs can only ever be a "best effort" attempt.

Reply:
Yes, there's at least one other existing test that relies on updating only dependent outputs, which I intentionally broke to find one while working on this ;)